src/abi/json-abi-format.md

```diff
@@ -35,52 +36,36 @@ The ABI of a contract is represented as a JSON object containing the following p
-  - `"components"`: an array of the components of a given type, if any, and `null` otherwise. Each component is a _type application_ represented as a JSON object that contains the following properties:
+- `"concreteTypes"`: an array describing all the _type concrete declarations_ used (or transitively used) in the ABI. Each _type concrete declaration_ is a JSON object that contains the following properties:
+  - `"type"`: a string representing the type, the `sha256` of this string generates the `concreteTypeId`.
+  - `"concreteTypeId"`: a unique string hash based ID. Generated as specified in [Hash Based Ids](./hash-based-ids.md).
+  - `"metadataTypeId"`: the the _type metadata declaration_ ID of this type, if the type metadata has components or is generic, otherwise non existent.
+  - `"typeArguments"`: an array of _type concrete declarations_ hash based IDs of the type parameters of the type, if the type is generic, otherwise non existent.
+- `"typesMetadata"`: an array describing all the _type metadata declarations_ used (or transitively used) in the ABI. Each _type metadata declaration_ is a JSON object that contains the following properties:
+  - `"type"`: a string representation of the _type metadata declaration_. The section [JSON ABI Format for Each Possible Metadata Type Declaration](#json-abi-format-for-each-possible-metadata-type-declaration) specifies the format for each possible type.
+  - `"metadataTypeId"`: a unique integer ID.
+  - `"components"`: an array of the components of a given type, if any, otherwise non existent. Each component is a _type application_ represented as a JSON object that contains the following properties:
     - `"name"`: the name of the component.
-    - `"type"`: the _type declaration_ hash based ID of the type of the component.
-    - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the component, if the type is generic, and `null` otherwise. Each _type argument_ is a _type application_ represented as a JSON object that contains the following properties:
-      - `"type"`: the _type declaration_ hash based ID of the type of the _type argument_.
-      - `"typeArguments"`: an array of the _type arguments_ used when applying the type of the _type argument_, if the type is generic, and `null` otherwise. The format of the elements of this array recursively follows the rules described in this section.
-  - `"typeParameters"`: an array of type hash based IDs of the _type parameters_ of the type, if the type is generic, and `null` otherwise. Each _type parameter_ is a type declaration and is represented as described in [Generic Type Parameter](#generic-type-parameter).
+    - `"typeId"`: the _type metadata declaration_ ID or _type concrete declaration_ hash based ID of the type of the component.
```

Comment:
This `typeId` is currently _always_ a string, even though it might point to a `metadataTypeId` which is an integer. If we were to make it a _number or string_, we'd know based on the value's type if it's a metadata type or a concrete type and we wouldn't have to do any string parsing. It's a minor change proposal, though, and one can live with the current spec + string parsing as well.